File: demos/.okareo/flows/summarization.ts
import { 
    Okareo, 
    RunTestProps, components,
    TestRunType, OpenAIModel,
    GenerationReporter, CheckOutputType, JSONReporter
} from "okareo-ts-sdk";
import * as core from '@actions/core';

import { prompts } from './prompts/meeting_summary';
import { CHECK_TYPE, register_checks } from './utils/check_utils';

const OKAREO_API_KEY = process.env.OKAREO_API_KEY;
const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
const UNIQUE_BUILD_ID = (process.env.DEMO_BUILD_ID || `local.${(Math.random() + 1).toString(36).substring(7)}`);

const PROJECT_NAME = "Global";
const MODEL_NAME = "Meeting Summarizer";

const required_checks: CHECK_TYPE[] = [
	{
		name: "demo.Summary.Length",
		description: "Return the length of the short_summary property from the JSON model response.",
		output_data_type: CheckOutputType.SCORE,
	},
	{
		name: "demo.Summary.Under256",
		description: "Pass if the property short_summary from the JSON model result has less than 256 characters.",
		output_data_type: CheckOutputType.PASS_FAIL,
	},
	{
		name:"demo.Summary.JSON",
		description: "Pass if the model result is JSON with the properties short_summary, actions, and attendee_list.",
		output_data_type: CheckOutputType.PASS_FAIL,
	},
	{
		name:"demo.Attendees.Length",
		description: "Return the length of the number of particpants in the attendee_list in the JSON model response.",
		output_data_type: CheckOutputType.SCORE,
	},
	{
		name:"demo.Actions.Length",
		description: "Return the length of the number of actions in the JSON model response.",
		output_data_type: CheckOutputType.SCORE,
	},
	{
		name:"demo.Tone.IsFriendly",
		description: "Use a model judgement to determine if the tone in the meeting is friendly (true).",
		prompt: "Only output True if the speakers in the following meeting are friendly, otherwise return False: {generation}",
		output_data_type: CheckOutputType.PASS_FAIL,
	},
];

const report_definition = {
	metrics_min: {
		"consistency": 4.0,
		"relevance": 4.0,
		"demo.Attendees.Length": 2,
		"demo.Actions.Length": 2,
	},
	metrics_max: {
		"demo.Summary.Length": 256,
	},
	pass_rate: {
		"demo.Summary.Under256": 0.75,
		"demo.Summary.JSON": 1,
		"demo.Tone.Friendly": 1,
	},
};

const main = async () => {
	try {
		const okareo = new Okareo({api_key:OKAREO_API_KEY});
		const pData: any[] = await okareo.getProjects();
		const project_id = pData.find(p => p.name === PROJECT_NAME)?.id;

		register_checks(okareo, project_id, required_checks);

		const meeting_scenario: any = await okareo.upload_scenario_set({
			name: "Meeting Bank Small Data Set",
			file_path: "./.okareo/flows/meetings.jsonl",
			project_id: project_id,
		});
		
		const model = await okareo.register_model({
		name: MODEL_NAME,
		tags: ["Demo", "Summaries", `Build:${UNIQUE_BUILD_ID}`],
		project_id: project_id,
		models: {
			type: "openai",
			model_id:"gpt-3.5-turbo",
			temperature:0.5,
			system_prompt_template:prompts.getSummarySystemPrompt(),
			user_prompt_template:prompts.getSummaryUserPrompt(),
		} as OpenAIModel,
		update: true,
		});
		
		const eval_run: components["schemas"]["TestRunItem"] = await model.run_test({
			model_api_key: OPENAI_API_KEY,
			name: `${MODEL_NAME} Eval ${UNIQUE_BUILD_ID}`,
			tags: ["Demo", "Summaries", `Build:${UNIQUE_BUILD_ID}`],
			project_id: project_id,
			scenario: meeting_scenario,
			calculate_metrics: true,
			type: TestRunType.NL_GENERATION,
			checks: [
				"consistency_summary",
				"relevance_summary",
				...required_checks.map(c => c.name),
			]
		} as RunTestProps);

		const reporter = new GenerationReporter({
				eval_run :eval_run, 
				...report_definition,
		});
		reporter.log();

        const reporter_output = new JSONReporter({
            eval_runs:[ eval_run ]
        });
        reporter_output.log();
		
		if (!reporter.pass) {
			// intentionally not blocking the build.
			console.log("The model did not pass the evaluation. Please review the results.");
			//core.setFailed("The model did not pass the evaluation. Please review the results.");
		}

	} catch (error) {
        // intentionally not blocking the build.
		core.setFailed("Failed: "+error);
	}
}
main();

File: demos/.okareo/flows/summarization.py
import os
from importlib import import_module
from prompts.meeting_summary import Prompts

from okareo import Okareo
from okareo.checks import CheckOutputType
from okareo.model_under_test import GenerationModel
from okareo_api_client.models.test_run_type import TestRunType
from okareo_api_client.models import EvaluatorSpecRequest

# Set up API keys from environment variables
OKAREO_API_KEY = os.environ['OKAREO_API_KEY']
ANTHROPIC_API_KEY = os.environ['ANTHROPIC_API_KEY']

# Define model IDs and names for different Claude 3 versions
opus_model_id = "claude-3-opus-20240229"
opus_model_name = "Claude 3 Opus"

sonnet_model_id = "claude-3-sonnet-20240229"
sonnet_model_name = "Claude 3 Sonnet"

haiku_model_id = "claude-3-haiku-20240307"
haiku_model_name = "Claude 3 Haiku"

# Select the model to use (currently set to Haiku)
model_id = opus_model_id
model_name = opus_model_name

# Define custom checks for evaluation
custom_checks = [
	{
		"name": "character_count",
		"description": "Return the length of model output in characters",
		"output_data_type": CheckOutputType.SCORE,
	},
	{
		"name": "under_350_characters",
		"description": "Pass if the property model output from the JSON model result has less than 350 characters.",
		"output_data_type": CheckOutputType.PASS_FAIL,
	},
	{
		"name": "word_count",
		"description": "Return the number of words in the model output",
		"output_data_type": CheckOutputType.SCORE,
	},
]

# Initialize Okareo client
okareo = Okareo(OKAREO_API_KEY)

# Upload scenario set for testing
scenario = okareo.upload_scenario_set(
    scenario_name="Meeting Bank Summaries (Test)",
    file_path='./.okareo/flows/Meeting_Bank_Summaries_test.jsonl'
)

# Register the model under test
mut = okareo.register_model(
	name=f"{model_name} (Summarization)",
	model=GenerationModel(
        model_id=model_id,
		temperature=0,
		system_prompt_template=Prompts.get_summary_system_prompt(),
		user_prompt_template=Prompts.get_summary_user_prompt(),
	),
    update=True,
)

# Get all existing checks
all_checks = okareo.get_all_checks()
all_check_names = [check.name for check in all_checks]

# Create or update custom checks
for check_info in custom_checks:
	# Skip generating checks that already exist
	if check_info['name'] in all_check_names:
		continue
	
	# Generate checks
	generated_test = okareo.generate_check(EvaluatorSpecRequest(
		description=check_info.get('description'),
		requires_scenario_input=True,
		requires_scenario_result=False,
		output_data_type=check_info.get('output_data_type').value
	)).generated_code
	file_path = f"./.okareo/flows/{check_info.get('name')}.py"
	with open(file_path, "w+") as file:
		file.write(generated_test)
	check_module = import_module(check_info.get('name'))
	Check = getattr(check_module, 'Check')
	okareo.create_or_update_check(
		name=check_info.get('name'),
		check=Check(),
		description=check_info.get('description'),
	)
	os.remove(file_path)

# Get the names of all custom checks
custom_check_names = [check_info['name'] for check_info in custom_checks]

# Run the evaluation
eval_run = mut.run_test(
    scenario=scenario,
    name=f"Summarization Run (test split) for {model_name}",
    test_run_type=TestRunType.NL_GENERATION,
    checks=['latency', 'consistency_summary'] + custom_check_names,
    api_key=ANTHROPIC_API_KEY
)

print(f'{eval_run.name} can be viewed at {eval_run.app_link}')